src/fsharp/ConstraintSolver.fs

```diff
@@ -705,77 +708,80 @@ let CheckWarnIfRigid (csenv:ConstraintSolverEnv) ty1 (r:Typar) ty =
+                    // We may need to make use of the equation when solving the constraints. 
+                    // Record a entry in the undo trace if one is provided 
+                    trace.Exec (fun () -> r.typar_solution <- Some ty) (fun () -> r.typar_solution <- None)
       
-      // Re-solve the other constraints associated with this type variable 
-      solveTypMeetsTyparConstraints csenv ndeep m2 trace ty r))
+                    (* dprintf "setting typar %d to type %s at %a\n" r.Stamp ((DebugPrint.showType ty)) outputRange m; *)
 
-    | _ -> failwith "SolveTyparEqualsType")
+                    // Only solve constraints if this is not an error var 
+                    if r.IsFromError then () 
+                    else
+                        // Check to see if this type variable is relevant to any trait constraints. 
+                        // If so, re-solve the relevant constraints. 
+                        if csenv.SolverState.ExtraCxs.ContainsKey r.Stamp then 
+                            do! RepeatWhileD ndeep (fun ndeep -> SolveRelevantMemberConstraintsForTypar csenv ndeep false trace r)
+                        // Re-solve the other constraints associated with this type variable 
+                        return! solveTypMeetsTyparConstraints csenv ndeep m2 trace ty r
+
+        | _ -> failwith "SolveTyparEqualsType"
+    }
     
 
 /// Apply the constraints on 'typar' to the type 'ty'
 and solveTypMeetsTyparConstraints (csenv:ConstraintSolverEnv) ndeep m2 trace ty (r: Typar) = 
     let g = csenv.g
-    // Propagate compat flex requirements from 'tp' to 'ty'
-    SolveTypIsCompatFlex csenv trace r.IsCompatFlex ty ++ (fun () -> 
-    // Propagate dynamic requirements from 'tp' to 'ty'
-    SolveTypDynamicReq csenv trace r.DynamicReq ty ++ (fun () -> 
-    // Propagate static requirements from 'tp' to 'ty' 
-    SolveTypStaticReq csenv trace r.StaticReq ty ++ (fun () -> 
+    trackErrors {
+        // Propagate compat flex requirements from 'tp' to 'ty'
+        do! SolveTypIsCompatFlex csenv trace r.IsCompatFlex ty
+        
+        // Propagate dynamic requirements from 'tp' to 'ty'
+        do! SolveTypDynamicReq csenv trace r.DynamicReq ty
+
+        // Propagate static requirements from 'tp' to 'ty' 
+        do! SolveTypStaticReq csenv trace r.StaticReq ty
     
-    // Solve constraints on 'tp' w.r.t. 'ty' 
-    r.Constraints |> IterateD (function
-      | TyparConstraint.DefaultsTo (priority, dty, m) -> 
-          if typeEquiv g ty dty then 
-              CompleteD
-          else
-              match tryDestTyparTy g ty with
-              | ValueNone -> CompleteD
-              | ValueSome destTypar ->
-                  AddConstraint csenv ndeep m2 trace destTypar (TyparConstraint.DefaultsTo(priority, dty, m))
+        // Solve constraints on 'tp' w.r.t. 'ty' 
+        return!
+            r.Constraints
+            |> IterateD (function
```

Comment:
Can you add a `For` method on `trackChanges` builder to replace `IterateD`?

Reply:
Hi!
Sure I can do this 😃